compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs

```diff
@@ -654,6 +655,64 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
         }
     }
 
+    /// Suggests moving redundant argument(s) of an associate function to the
+    /// trait it belongs to.
+    ///
+    /// ```compile_fail
+    /// Into::into::<Option<_>>(42) // suggests considering `Into::<Option<_>>::into(42)`
+    /// ```
+    fn suggest_moving_args(&self, err: &mut Diagnostic) {
+        if let Some(trait_) = self.tcx.trait_of_item(self.def_id) {
+            // HACK(hkmatsumoto): Ugly way to tell "<trait>::<assoc fn>()" from "x.<assoc fn>()";
+            // we don't care the latter (for now).
+            if self.path_segment.res == Some(hir::def::Res::Err) {
+                return;
+            }
+
+            // Say, if the assoc fn takes `A`, `B` and `C` as generic arguments while expecting 1
+            // argument, and its trait expects 2 arguments. It is hard to "split" them right as
+            // there are too many cases to handle: `A` `B` | `C`, `A` `B` | `C`, `A` `C` | `B`, ...
```

Comment:
I think this comment could be a little bit clearer, it took me a little bit to understand what it was trying to communicate, perhaps:

```suggestion
            // Skip suggestion when the associated function is itself generic, it is unclear
            // how to split the provided parameters between those to suggest to the trait and
            // those to remain on the associated type.
```

Reply:
This is more straight-to-the-point, and I like this. English is harder than any programming language I came across ;-)